src/metaopt/utils/__init__.py

```diff
@@ -19,16 +21,84 @@ def __init__(cls, name, bases, dictionary):
         super(SingletonType, cls).__init__(name, bases, dictionary)
         cls.instance = None
 
-    def __call__(cls, *args, **kwds):
+    def __call__(cls, *args, **kwargs):
         """Create an object if does not already exist, otherwise return what there is."""
         if cls.instance is None:
-            cls.args = args
-            cls.kwds = kwds
-            cls.instance = super(SingletonType, cls).__call__(*args, **kwds)
+            cls.instance = super(SingletonType, cls).__call__(*args, **kwargs)
+        elif args or kwargs:
+            raise ValueError("A singleton instance has already been instantiated.")
         return cls.instance
 
 
 class AbstractSingletonType(SingletonType, ABCMeta):
     """This will create singleton base classes, that need to be subclassed and implemented."""
 
     pass
+
+
+class Factory(type):
+    """Instantiate appropriate wrapper for the infrastructure based on input
+    argument, ``of_type``.
+
+    Attributes
+    ----------
+    types : list of subclasses of ``cls.__base__``
+       Updated to contain all possible implementations currently. Check out code.
+    typenames : list of str
+       Names of implemented wrapper classes, correspond to possible ``of_type``
+       values.
+
+    """
+
+    def __init__(cls, names, bases, dictionary):
+        """Search in directory for attribute names subclassing `bases[0]`"""
+        super(Factory, cls).__init__(names, bases, dictionary)
+
+        modules = []
+        base = import_module(cls.__base__.__module__)
+        py_files = map(lambda x: '.' + x.split('.')[0].split('/')[-1],
```

Comment:
That will be a painful to test. Why not depending on imports for making available the subclasses? It makes sense to import all classes of a given type in the same ```__init__.py``` file, ```database/__init__.py``` for instance. Also, it makes it easier to drop subclasses if they have unmet dependencies for instance. That way Database(of_type="PostgreSQL") would not be possible if the dependencies are unmet (it's not imported). Maybe dropping them for that is not a good idea but it's an example of how more flexible it is. The downside is making obligatory imports in specific ```__init__.py``` but that is a common practice anyway.

Reply:
I did it like because if it's pain to append a class name to a `types` list, then I assume it would be also a pain to require importing every new class to `__init__.py`. This works. You can checks the test code. It assumes that all implementations lie in files that reside in the same directory as the base implementation. What else should be the case, in order to be considered that it is tested properly?

(Not too much) Later this will be augmented with searching through the `pkg_resources` for registered `entry_points` also, in order to include the case where the interface is exported as a plugin (I consider the database interface to be an internal though).

Reply:
I was worried that we would try looking all around so there could be many corner cases to test, but what you explain here is simple. Let's try it that way.